File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

typedef struct date {
    int year;
    int month;
    int day;
}DATE;

bool isValid(DATE today, DATE privacy) {
    if (privacy.year < today.year)
        return true;
    if (privacy.year > today.year)
        return false;
    if (privacy.month < today.month)
        return true;
    if (privacy.month > today.month)
        return false;
    if (privacy.day < today.day)
        return true;
    return false;
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    DATE Today = {stoi(today.substr(0, 4)), stoi(today.substr(5, 2)), stoi(today.substr(8, 2))};
    map<char, int> types;
    for (int i = 0; i < terms.size(); i++)
        types.insert({terms[i][0], atoi(&terms[i][2])});
    vector<int> answer;
    for (int i = 0; i < privacies.size(); i++) {
        int p_year = stoi(privacies[i].substr(0, 4));
        int p_month = stoi(privacies[i].substr(5, 2));
        int p_day = stoi(privacies[i].substr(8, 2));
        char type = privacies[i][11];
        p_month += types[type];
        p_year += (p_month-1) / 12;
        p_month -= (p_month-1)/12*12;
        p_day--;
        if (p_day < 1) {
            p_day = 28;
            p_month--;
            if (p_month < 1) {
                p_month = 12;
                p_year--;
            }
        }
        if (isValid(Today, {p_year, p_month, p_day}))
            answer.emplace_back(i+1);
    }
    return answer;
}

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.swift
import Foundation

struct MyDate {
    var year: Int
    var month: Int
    var day: Int
    var days: Int
    
    init(_ date:String) {
        let date = Array(date)
        year = Int(String(date[...3]))!
        month = Int(String(date[5...6]))!
        day = Int(String(date[8...9]))!
        days = year*12*28 + month*28 + day
    }
}

func solution(_ today:String, _ terms:[String], _ privacies:[String]) -> [Int] {
    let today = MyDate(today)
    var types: [String: Int] = [:]
    for i in 0..<terms.count {
        var n = String(terms[i].suffix(terms[i].count-2))
        types[String(terms[i].prefix(1))] = Int(n)!
    }
    
    var answer = [Int]()
    for i in 0..<privacies.count {
        let privacy = MyDate(privacies[i])
        let type = String(privacies[i].suffix(1))
        if privacy.days + types[type]!*28 - 1 < today.days { answer.append(i+1) }
    }
    return answer
}